bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/SortedLedgerStorageCheckpointTest.java

```diff
@@ -0,0 +1,231 @@
+            }
+        });
+        this.storage.initialize(
+            conf,
+            mock(LedgerManager.class),
+            ledgerDirsManager,
+            ledgerDirsManager,
+            checkpointSrc,
+            checkpointer,
+            NullStatsLogger.INSTANCE);
+    }
+
+    @After
+    @Override
+    public void tearDown() throws Exception {
+        if (null != storage) {
+            storage.shutdown();
+        }
+        super.tearDown();
+    }
+
+    ByteBuf prepareEntry(long ledgerId, long entryId) {
+        ByteBuf entry = Unpooled.buffer(4 * Long.BYTES);
+        // ledger id, entry id, lac
+        entry.writeLong(ledgerId);
+        entry.writeLong(entryId);
+        entry.writeLong(entryId - 1);
+        // data
+        entry.writeLong(entryId);
+        return entry;
+    }
+
+    @Test
+    public void testCheckpoint() throws Exception {
+        // memory table holds the first checkpoint, but it is not completed yet.
+        Checkpoint memtableCp = storage.memTable.kvmap.cp;
+        assertEquals(new TestCheckpoint(0), memtableCp);
+
+        // write entries into ledger storage
+        long lid = System.currentTimeMillis();
+        storage.setMasterKey(lid, new byte[0]);
+        for (int i = 0; i < 20; i++) {
+            storage.addEntry(prepareEntry(lid, i));
+        }
+        // simulate journal persists the entries in journal;
+        checkpointSrc.advanceOffset(100);
+
+        // memory table holds the first checkpoint, but it is not completed yet.
+        memtableCp = storage.memTable.kvmap.cp;
+        assertEquals(new TestCheckpoint(0), memtableCp);
+
+        // trigger a memtable flush
+        storage.onSizeLimitReached(checkpointSrc.newCheckpoint());
+        // wait for checkpoint to complete
+        checkpoints.poll(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+        assertEquals(new TestCheckpoint(100), storage.memTable.kvmap.cp);
+        assertEquals(0, storage.memTable.kvmap.size());
+    }
+
+    @Test
+    public void testCheckpointAfterEntryLogRotated() throws Exception {
+        // memory table holds the first checkpoint, but it is not completed yet.
+        Checkpoint memtableCp = storage.memTable.kvmap.cp;
+        assertEquals(new TestCheckpoint(0), memtableCp);
+
+        // write entries into ledger storage
+        long lid = System.currentTimeMillis();
+        storage.setMasterKey(lid, new byte[0]);
+        for (int i = 0; i < 20; i++) {
+            storage.addEntry(prepareEntry(lid, i));
+        }
+        // simulate journal persists the entries in journal;
+        checkpointSrc.advanceOffset(100);
+
+        // memory table holds the first checkpoint, but it is not completed yet.
+        memtableCp = storage.memTable.kvmap.cp;
+        assertEquals(new TestCheckpoint(0), memtableCp);
+        assertEquals(20, storage.memTable.kvmap.size());
+
+        final CountDownLatch readyLatch = new CountDownLatch(1);
```

Comment:
What is the purpose of this?

Reply:
the purpose is to make sure we roll the log before checkpointing kicks in. it ensures the test case can be executed in a deterministic way.

Reply:
How does this work though? Is there already a runnable on the executor, so by adding another, we know when the runnable has completed? A comment around this would be very useful for anyone else coming back to this test in a couple of months.

Reply:
- rollLog simulates a entry log file rolling (for the use case that Charan reported)
- after this change, an underneath entry log rolling will not trigger any checkpoint, to ensure that there is not checkpoint happening, this test case insert a blocking runnable. so we can collect the right log id at line 209 and line 210 for the assertion at the end.

will add comments.

Reply:
👍

Reply:
@sijie This is ready to merge once that comment is added.

Reply:
@sijie can you add this comment?

Reply:
will add the comments after @reddycharan or @jvrao reviewed this.